File: src/main/java/com/adventofcode/aoc2017/AoC012017.java
package com.adventofcode.aoc2017;

import com.adventofcode.Solution;

import java.util.function.IntUnaryOperator;
import java.util.stream.Stream;

import static com.adventofcode.utils.Utils.*;

class AoC012017 implements Solution {

    private static String solve(final String s, final IntUnaryOperator computeNext) {
        int res = 0;
        for (int i = 0; i < s.length(); i++) {
            //compute where is the value to check
            final int next = computeNext.applyAsInt(i);
            final char toCheck = s.charAt(next);
            final char curr = s.charAt(i);
            if (curr == toCheck) {
                res += charToInt(curr);
            }
        }
        return itoa(res);
    }

    public String solveFirstPart(final Stream<String> input) {
        final String s = getFirstString(input);
        return solve(s, i -> (i + 1) % s.length());
    }

    public String solveSecondPart(final Stream<String> input) {
        final String s = getFirstString(input);
        return solve(s, i -> (i + s.length() / 2) % s.length());
    }
}

File: src/main/kotlin/com/adventofcode/aoc2017/AoC012017.kt
package com.adventofcode.aoc2017

import com.adventofcode.Solution
import com.adventofcode.utils.Utils.*
import java.util.stream.Stream

class AoC012017 : Solution {

    private fun solve(s: String, computeNext: (Int) -> Int): String {
        var res = 0
        for (i in 0 until s.length) {
            //compute where is the value to check
            val next = computeNext.invoke(i)
            val toCheck = s[next]
            val curr = s[i]
            if (curr == toCheck) {
                res += charToInt(curr)
            }
        }
        return itoa(res.toLong())
    }

    override fun solveFirstPart(input: Stream<String>): String {
        val s = getFirstString(input)
        return solve(s) { i -> (i + 1) % s.length }
    }

    override fun solveSecondPart(input: Stream<String>): String {
        val s = getFirstString(input)
        return solve(s) { i -> (i + s.length / 2) % s.length }
    }
}
